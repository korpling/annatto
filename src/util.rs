use crate::{
    error::{AnnattoError, Result},
    importer::Importer,
};
use graphannis::{
    model::AnnotationComponentType,
    update::{GraphUpdate, UpdateEvent},
    AnnotationGraph,
};
use graphannis_core::graph::ANNIS_NS;
use itertools::Itertools;
use std::{
    collections::BTreeMap,
    fs::File,
    io::{BufWriter, Write},
    path::Path,
};

pub mod graphupdate;

fn event_to_string(update_event: &UpdateEvent) -> Result<String> {
    Ok(format!("{:?}", update_event))
}

pub fn write_to_file(updates: &GraphUpdate, path: &std::path::Path) -> Result<()> {
    let mut file = File::create(path)?;
    let it = updates.iter()?;
    for update_event in it {
        let event_tuple = update_event?;
        let event_string = event_to_string(&event_tuple.1)?;
        file.write_all(event_string.as_bytes())?;
        file.write_all(b"\n")?;
    }
    Ok(())
}

pub fn insert_corpus_nodes_from_path(
    update: &mut GraphUpdate,
    root_path: &Path,
    document_path: &Path,
) -> Result<String> {
    let clean_path = normpath::BasePath::new(document_path)?;
    let clean_root_path = normpath::BasePath::new(root_path)?;
    let norm_path = normpath::BasePath::normalize(&clean_path)?;
    let norm_root_path = normpath::BasePath::normalize(&clean_root_path)?;
    let root_path_len = norm_root_path.components().count() - 1;
    let mut full_path = String::new();
    for c in &norm_path.components().collect_vec()[root_path_len..] {
        let parent = full_path.clone();
        if !full_path.is_empty() {
            full_path += "/";
        }
        full_path += &c.as_os_str().to_string_lossy();
        update.add_event(UpdateEvent::AddNode {
            node_name: full_path.to_string(),
            node_type: "corpus".to_string(),
        })?;
        if !parent.is_empty() {
            update.add_event(UpdateEvent::AddEdge {
                source_node: full_path.to_string(),
                target_node: parent,
                layer: ANNIS_NS.to_string(),
                component_type: AnnotationComponentType::PartOf.to_string(),
                component_name: "".to_string(),
            })?;
        }
    }
    Ok(full_path)
}

pub fn import_as_graphml_string<I, P>(
    importer: I,
    path: P,
    properties: BTreeMap<String, String>,
    graph_configuration: Option<&str>,
) -> Result<String>
where
    I: Importer,
    P: AsRef<Path>,
{
    let mut u = importer
        .import_corpus(path.as_ref(), &properties, None)
        .map_err(|e| AnnattoError::Import {
            reason: e.to_string(),
            importer: importer.module_name().to_string(),
            path: path.as_ref().to_path_buf(),
        })?;
    let mut g = AnnotationGraph::with_default_graphstorages(false)?;
    g.apply_update(&mut u, |_| {})?;

    let mut buf = BufWriter::new(Vec::new());
    graphannis_core::graph::serialization::graphml::export(
        &g,
        graph_configuration,
        &mut buf,
        |_| {},
    )?;
    let bytes = buf.into_inner()?;
    let actual = String::from_utf8(bytes)?;

    Ok(actual)
}

#[cfg(test)]
mod tests {
    use std::{env::current_dir, path::Path};

    use graphannis::update::GraphUpdate;

    const TEST_PATH: &str = "tests/data/import/xlsx";

    fn test_insert_corpus_nodes_from_path(
        absolute: bool,
    ) -> Result<(), Box<dyn std::error::Error>> {
        let sys_path = current_dir()?;
        let p = Path::new(TEST_PATH);
        let test_path = if absolute {
            sys_path.join(p)
        } else {
            p.to_path_buf()
        };
        let mut u = GraphUpdate::default();
        let parent_path = test_path.parent().unwrap().parent().unwrap();

        let r = super::insert_corpus_nodes_from_path(&mut u, &parent_path, test_path.as_path());
        assert!(r.is_ok(), "Not okay: {:?}", r.err());
        let doc_path = r?;
        assert_eq!(doc_path, "data/import/xlsx");
        Ok(())
    }

    #[test]
    fn test_insert_corpus_nodes_from_path_relative() {
        let r = test_insert_corpus_nodes_from_path(false);
        assert!(r.is_ok(), "Not okay: {:?}", r.err());
    }

    #[test]
    fn test_insert_corpus_nodes_from_path_absolute() {
        let r = test_insert_corpus_nodes_from_path(true);
        assert!(r.is_ok(), "Not okay: {:?}", r.err());
    }
}
