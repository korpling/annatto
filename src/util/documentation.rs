use crate::ModuleConfiguration;
use facet::{Type, UserType, Variant};
use itertools::Itertools;

/// Gets a clean documentation string for a [`Shape`].
/// It removes any whitespace at the start and unescapes backlashes.
pub fn clean_string(raw: &[&str]) -> String {
    raw.iter()
        .map(|line| line.trim_start())
        .map(unescape_backslash)
        .join("\n")
}

/// Replaces characters that are escaped with a backslash with the actual
/// character. Only unescapes the characters that are escaped by facet.
fn unescape_backslash(val: &str) -> String {
    let mut chars = val.chars().peekable();
    let mut unescaped = String::new();

    loop {
        match chars.next() {
            None => break,
            Some(c) => {
                let escaped_char = if c == '\\' {
                    if let Some(escaped_char) = chars.peek() {
                        let escaped_char = *escaped_char;
                        match escaped_char {
                            _ if escaped_char == '\\'
                                || escaped_char == '\''
                                || escaped_char == '`'
                                || escaped_char == '$' =>
                            {
                                Some(escaped_char)
                            }
                            'n' => Some('\n'),
                            'r' => Some('\r'),
                            't' => Some('\t'),
                            _ => None,
                        }
                    } else {
                        None
                    }
                } else {
                    None
                };
                if let Some(escaped_char) = escaped_char {
                    unescaped.push(escaped_char);
                    // skip the escaped character instead of outputting it again
                    chars.next();
                } else {
                    unescaped.push(c);
                };
            }
        }
    }

    unescaped
}

pub struct ModuleInfo {
    pub name: String,
    pub doc: String,
    pub configs: Vec<ModuleConfiguration>,
}

impl From<&Variant> for ModuleInfo {
    fn from(module: &Variant) -> Self {
        // The name of the module is taken from the wrapper enum
        let module_name = module.name.to_lowercase();
        // Get the inner type wrapped by the graph operations enum and use
        // its documentation and fields
        let mut result = Self {
            name: module_name,
            doc: "".to_string(),
            configs: Vec::new(),
        };
        if let Some(inner_field) = module.data.fields.first().map(|m| m.shape())
            && let Type::User(module_type) = inner_field.ty
            && let UserType::Struct(module_impl) = module_type
        {
            result.doc = clean_string(inner_field.doc);

            result.configs = module_impl
                .fields
                .iter()
                .map(|f| ModuleConfiguration {
                    name: f.name.to_lowercase(),
                    description: clean_string(f.doc),
                })
                .collect();
        }
        result
    }
}

#[cfg(test)]
mod tests {

    use super::*;

    #[test]
    fn test_unescape_backslash() {
        assert_eq!(unescape_backslash("ab\\$c"), "ab$c");
        assert_eq!(unescape_backslash("ab\\\\cd\\\\"), "ab\\cd\\",);
        assert_eq!(unescape_backslash("ab\\'cd\\te"), "ab'cd\te");
        assert_eq!(unescape_backslash("a\\n"), "a\n");
    }
}
