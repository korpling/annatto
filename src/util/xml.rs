use std::io::BufRead;

use anyhow::{anyhow, Result};
use quick_xml::{
    events::{attributes::Attributes, Event},
    Reader,
};

/// Extract an attribute for an XML element by the namespace and name.
pub(crate) fn get_attribute_by_qname<'a>(
    attribute_list: Attributes<'a>,
    namespace: &str,
    name: &str,
) -> Result<Option<String>> {
    for att in attribute_list {
        let att = att?;
        if let Some(prefix) = att.key.prefix() {
            if prefix.as_ref() == namespace.as_bytes()
                && att.key.local_name().as_ref() == name.as_bytes()
            {
                let value = String::from_utf8_lossy(&att.value).to_string();
                return Ok(Some(value));
            }
        }
    }
    Ok(None)
}

/// Extract an attribute for an XML element by the name.
pub(crate) fn get_attribute_by_local_name<'a>(
    attribute_list: Attributes<'a>,
    name: &str,
) -> Result<Option<String>> {
    for att in attribute_list {
        let att = att?;
        if att.key.local_name().as_ref() == name.as_bytes() {
            let value = String::from_utf8_lossy(&att.value).to_string();
            return Ok(Some(value));
        }
    }
    Ok(None)
}

/// Read the next event. Will fail if the next event is not a start tag and does
/// not have the given name. All non tag elements (XML declaration, text nodes,
/// comments, ...) are ignored and skipped.
pub(crate) fn consume_start_tag_with_name<R>(reader: &mut Reader<R>, name: &str) -> Result<()>
where
    R: BufRead,
{
    let mut buf = Vec::new();
    loop {
        let event = reader.read_event_into(&mut buf)?;
        let result = match event {
            Event::Start(tag) => {
                if tag.local_name().as_ref() == name.as_bytes() {
                    Ok(())
                } else {
                    Err(anyhow!(
                        "Expected <{name}> but got <{}>",
                        String::from_utf8_lossy(tag.local_name().as_ref())
                    ))
                }
            }
            Event::End(_) => Err(anyhow!(
                "Expected \"<{name}>\" but got closing tag instead."
            )),
            Event::Empty(_) => Err(anyhow!("Expected \"<{name}>\" but got empty tag instead.")),

            Event::Comment(_)
            | Event::Decl(_)
            | Event::PI(_)
            | Event::DocType(_)
            | Event::CData(_)
            | Event::Text(_) => continue,
            Event::Eof => Err(anyhow!(
                "Expected <{name} but the file is already at its end."
            )),
        };
        return result;
    }
}
