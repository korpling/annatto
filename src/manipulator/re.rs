use graphannis::{AnnotationGraph, update::{GraphUpdate, UpdateEvent}};
use graphannis_core::{annostorage::ValueSearch, graph::NODE_NAME_KEY};
use graphannis_core::util::split_qname;
use itertools::Itertools;

use crate::{Manipulator, Module};

pub struct Rename {}
pub struct Remove {}

const RENAME_ID: &str = "rename";
const REMOVE_ID: &str = "remove";

impl Default for Rename {
    fn default() -> Self {
        Rename {}
    }
}

impl Default for Remove {
    fn default() -> Self {
        Remove {}
    }
}

impl Module for Rename {
    fn module_name(&self) -> &str {
        RENAME_ID
    }
}

impl Module for Remove {
    fn module_name(&self) -> &str {
        REMOVE_ID
    }
}

const PROP_NODE_ANNOS: &str = "node.annos";
const PROP_NODE_NAMES: &str = "node.names";
const PROP_EDGE_ANNOS: &str = "edge.annos";
const PROPVAL_SEP: &str = ",";

fn remove_nodes(graph: &mut AnnotationGraph, names: Vec<&str>) -> Result<(), Box<dyn std::error::Error>> {
    let mut update = GraphUpdate::default();
    for name in names {
        update.add_event(UpdateEvent::DeleteNode { node_name: name.to_string() })?;
    }
    graph.apply_update(&mut update, |_| {})?;
    Ok(())
}

fn remove_node_annos(graph: &mut AnnotationGraph, names: Vec<(Option<&str>, &str)>) -> Result<(), Box<dyn std::error::Error>> {
    let mut update = GraphUpdate::default();
    let annos = graph.get_node_annos();
    for (ns, name) in names.into_iter() {
        for r in annos.exact_anno_search(ns, name, ValueSearch::Any) {
            let m = r?;
            let node_name = annos.get_value_for_item(&m.node, &NODE_NAME_KEY)?.unwrap();
            let del_ns = match ns {
                None => "".to_string(),
                Some(v) => v.to_string()
            };
            update.add_event(UpdateEvent::DeleteNodeLabel { node_name: node_name.to_string(), 
                                                            anno_ns: del_ns, 
                                                            anno_name: name.to_string() })?;
        }
    }
    graph.apply_update(&mut update, |_| {})?;
    Ok(())
}

fn remove_edge_annos(graph: &mut AnnotationGraph, names: Vec<(Option<&str>, &str)>) -> Result<(), Box<dyn std::error::Error>> {
    let mut update = GraphUpdate::default();
    let node_annos = graph.get_node_annos();
    for (ns, name) in names {
        for component in graph.get_all_components(None, None) {
            let component_storage = graph.get_graphstorage(&component).unwrap();            
            let edge_annos = component_storage.get_anno_storage();           
            for r in edge_annos.exact_anno_search(ns, name, ValueSearch::Any) {
                let m = r?;                
                let source_node = m.node;
                let source_node_name = node_annos.get_value_for_item(&source_node, &NODE_NAME_KEY)?.unwrap();
                for out_edge_opt in component_storage.get_outgoing_edges(source_node) {                    
                    let target_node = out_edge_opt?;
                    let target_node_name = node_annos.get_value_for_item(&target_node, &NODE_NAME_KEY)?.unwrap();
                    update.add_event(UpdateEvent::DeleteEdgeLabel { source_node: source_node_name.to_string(), 
                                                                    target_node: target_node_name.to_string(), 
                                                                    layer: component.layer.to_string(), 
                                                                    component_type: component.get_type().to_string(), 
                                                                    component_name: component.name.to_string(), 
                                                                    anno_ns: m.anno_key.ns.to_string(), 
                                                                    anno_name: name.to_string() })?;
                }
            }
        }
    }
    graph.apply_update(&mut update, |_| {})?;
    Ok(())
}

impl Manipulator for Remove {
    fn manipulate_corpus(
        &self,
        graph: &mut graphannis::AnnotationGraph,
        properties: &std::collections::BTreeMap<String, String>,
        _tx: Option<crate::workflow::StatusSender>,
    ) -> Result<(), Box<dyn std::error::Error>> {
        if let Some(node_name_s ) = properties.get(&PROP_NODE_NAMES.to_string()) {
            let node_names = node_name_s.split(PROPVAL_SEP).collect_vec();
            remove_nodes(graph, node_names)?;
        }
        if let Some(node_name_s ) = properties.get(&PROP_NODE_ANNOS.to_string()) {
            let node_annos = node_name_s.split(PROPVAL_SEP).map(|s| split_qname(s)).collect_vec();
            remove_node_annos(graph, node_annos)?;
        }
        if let Some(edge_name_s) = properties.get(&PROP_EDGE_ANNOS.to_string()) {
            let edge_annos = edge_name_s.split(PROPVAL_SEP).map(|s| split_qname(s)).collect_vec();
            remove_edge_annos(graph, edge_annos)?;
        }
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use std::collections::{BTreeMap, HashSet};
    use std::env::temp_dir;

    use crate::Result;
    use crate::manipulator::Manipulator;
    use crate::manipulator::re::Remove;

    use graphannis::{AnnotationGraph,CorpusStorage};
    use graphannis::corpusstorage::{QueryLanguage,ResultOrder,SearchQuery};
    use graphannis_core::annostorage::ValueSearch;
    use graphannis::model::AnnotationComponentType;
    use graphannis::update::{GraphUpdate,UpdateEvent};
    use graphannis_core::graph::{ANNIS_NS, NODE_TYPE_KEY, NODE_NAME_KEY};
    use itertools::Itertools;
    use tempfile::{tempfile, tempdir_in};

    #[test]
    fn test_remove_in_mem() {
        let r = core_test(false); 
        assert_eq!(r.is_ok(), true, "Probing core test result {:?}", r);
    }

    #[test]
    fn test_remove_on_disk() {
        let r = core_test(true); 
        assert_eq!(r.is_ok(), true, "Probing core test result {:?}", r);
    }

    fn core_test(on_disk: bool) -> Result<()> {
        let mut g = input_graph(on_disk)?;
        let mut properties = BTreeMap::new();
        properties.insert("edge.annos".to_string(), "deprel".to_string());
        properties.insert("node.annos".to_string(), "pos".to_string());
        let remove = Remove::default();
        let result = remove.manipulate_corpus(&mut g, &properties, None);
        assert_eq!(result.is_ok(), true, "Probing merge result {:?}", &result);
        let mut e_g = expected_output_graph(on_disk)?;
        // corpus nodes
        let e_corpus_nodes: HashSet<String> = e_g.get_node_annos()
                                        .exact_anno_search(Some(&NODE_TYPE_KEY.ns), &NODE_TYPE_KEY.name, ValueSearch::Some("corpus"))
                                        .into_iter()
                                        .map(|r| r.unwrap().node)
                                        .map(|id_| e_g.get_node_annos().get_value_for_item(&id_, &NODE_NAME_KEY).unwrap().unwrap().to_string())
                                        .collect();
        let g_corpus_nodes: HashSet<String> = g.get_node_annos()
                                        .exact_anno_search(Some(&NODE_TYPE_KEY.ns), &NODE_TYPE_KEY.name, ValueSearch::Some("corpus"))
                                        .into_iter()
                                        .map(|r| r.unwrap().node)
                                        .map(|id_| g.get_node_annos().get_value_for_item(&id_, &NODE_NAME_KEY).unwrap().unwrap().to_string())
                                        .collect();
        assert_eq!(e_corpus_nodes, g_corpus_nodes);  //TODO clarify: Delegate or assertion?
        // test by components
        let e_c_list = e_g.get_all_components(None, None)
                                            .into_iter()
                                            .filter(|c| e_g.get_graphstorage(c).unwrap().source_nodes().count() > 0)
                                            .collect_vec();
        let g_c_list = g.get_all_components(None, None)
                                            .into_iter()
                                            .filter(|c| g.get_graphstorage(c).unwrap().source_nodes().count() > 0)  // graph might contain empty components after merge
                                            .collect_vec();
        assert_eq!(e_c_list.len(), g_c_list.len(), "components expected:\n{:?};\ncomponents are:\n{:?}", &e_c_list, &g_c_list);
        for c in e_c_list {
            let candidates = g.get_all_components(Some(c.get_type()), Some(c.name.as_str()));
            assert_eq!(candidates.len(), 1);
            let c_o  = candidates.get(0);
            assert_eq!(&c, c_o.unwrap());
        }
        // test with queries
        let queries = [
            "tok",
            "text",
            "lemma",
            "pos",
            "node ->dep node",
            "node ->dep[deprel=/.+/] node"
        ];
        let corpus_name = "current";
        let tmp_dir_e = tempdir_in(temp_dir())?;
        let tmp_dir_g = tempdir_in(temp_dir())?;        
        e_g.save_to(&tmp_dir_e.path().join(corpus_name))?;
        g.save_to(&tmp_dir_g.path().join(corpus_name))?;        
        let cs_e = CorpusStorage::with_auto_cache_size(&tmp_dir_e.path(), true)?;
        let cs_g = CorpusStorage::with_auto_cache_size(&tmp_dir_g.path(), true)?;
        for query_s in queries {
            let query = SearchQuery {
                corpus_names: &[corpus_name],
                query: query_s,
                query_language: QueryLanguage::AQL,
                timeout: None
            };
            let matches_e = cs_e.find(query.clone(), 0, None, ResultOrder::Normal)?;
            let matches_g = cs_g.find(query, 0, None, ResultOrder::Normal)?;
            assert_eq!(matches_e.len(), matches_g.len(), "Failed with query: {}", query_s);
            for (m_e, m_g) in matches_e.into_iter().zip(matches_g.into_iter()) {
                assert_eq!(m_e, m_g);
            }
        }
        Ok(())
    }

    #[test]
    fn test_export_mem() {
        let export = export_test(false);
        assert_eq!(export.is_ok(), true, "Export test ends with Err: {:?}", &export);
    }

    #[test]
    fn test_export_disk() {
        let export = export_test(true);
        assert_eq!(export.is_ok(), true, "Export test ends with Err: {:?}", &export);
    }

    fn export_test(on_disk: bool) -> Result<()> {
        let mut g = input_graph(on_disk)?;
        let mut properties = BTreeMap::new();
        properties.insert("edge.annos".to_string(), "deprel".to_string());
        properties.insert("node.annos".to_string(), "pos".to_string());
        let remover = Remove::default();
        assert_eq!(remover.manipulate_corpus(&mut g, &properties, None).is_ok(), true);
        let tmp_file = tempfile()?;
        let export = graphannis_core::graph::serialization::graphml::export(&g, None, tmp_file, |_| {});
        assert_eq!(export.is_ok(), true, "Export fails: {:?}", &export);
        Ok(())
    }

    fn input_graph(on_disk: bool) -> Result<AnnotationGraph> {
        let mut g = AnnotationGraph::new(on_disk)?;
        let mut u = GraphUpdate::default();
        u.add_event(UpdateEvent::AddNode { node_name: "root".to_string(), node_type: "corpus".to_string() })?;        
        u.add_event(UpdateEvent::AddNode { node_name: "root/b".to_string(), node_type: "corpus".to_string() })?;
        u.add_event(UpdateEvent::AddEdge { source_node: "root/b".to_string(), 
                                           target_node: "root".to_string(), 
                                           layer: ANNIS_NS.to_string(), 
                                           component_type: AnnotationComponentType::PartOf.to_string(), 
                                           component_name: "".to_string() })?;
        u.add_event(UpdateEvent::AddNode { node_name: "root/b/doc".to_string(), node_type: "corpus".to_string() })?;
        u.add_event(UpdateEvent::AddEdge { source_node: "root/b/doc".to_string(), 
                                           target_node: "root/b".to_string(), 
                                           layer: ANNIS_NS.to_string(), 
                                           component_type: AnnotationComponentType::PartOf.to_string(), 
                                           component_name: "".to_string() })?;
        for (ii, (txt, lemma_label, pos_label)) in [("I", "I", "PRON"),
                                                  ("am", "be", "VERB"), 
                                                  ("in", "in", "ADP"), 
                                                  ("Berlin", "Berlin", "PROPN")].iter().enumerate() {
            let i = ii + 1;
            let name = format!("root/b/doc#t{}", i);
            u.add_event(UpdateEvent::AddNode { node_name: name.to_string(), node_type: "node".to_string() })?;
            u.add_event(UpdateEvent::AddNodeLabel { node_name: name.to_string(), 
                                                    anno_ns: ANNIS_NS.to_string(), 
                                                    anno_name: "tok".to_string(), 
                                                    anno_value: txt.to_string() })?;
            u.add_event(UpdateEvent::AddNodeLabel { node_name: name.to_string(), 
                                                    anno_ns: "".to_string(), 
                                                    anno_name: "text".to_string(), 
                                                    anno_value: txt.to_string() })?;
            u.add_event(UpdateEvent::AddNodeLabel { node_name: name.to_string(), 
                                                    anno_ns: "".to_string(), 
                                                    anno_name: "lemma".to_string(), 
                                                    anno_value: lemma_label.to_string() })?;
            u.add_event(UpdateEvent::AddNodeLabel { node_name: name.to_string(), 
                                                    anno_ns: "".to_string(), 
                                                    anno_name: "pos".to_string(), 
                                                    anno_value: pos_label.to_string() })?;
            if i > 1 {
                u.add_event(UpdateEvent::AddEdge { source_node: format!("root/b/doc#t{}", i - 1), 
                                                   target_node: name.to_string(), 
                                                   layer: ANNIS_NS.to_string(), 
                                                   component_type: AnnotationComponentType::Ordering.to_string(), 
                                                   component_name: "".to_string() })?;
                u.add_event(UpdateEvent::AddEdge { source_node: format!("root/b/doc#t{}", i - 1), 
                                                   target_node: name.to_string(), 
                                                   layer: ANNIS_NS.to_string(), 
                                                   component_type: AnnotationComponentType::Ordering.to_string(), 
                                                   component_name: "text".to_string() })?;
            }
        }
        let dep_layer_name = "syntax";
        let dep_comp_name = "dep";
        let deprel_name = "deprel";
        for (source, target, label) in [(2, 1, "subj"),
                                      (2, 3, "comp:pred"),
                                      (3, 4, "comp:obj")].iter() {
            let source_name = format!("root/b/doc#t{}", source);
            let target_name = format!("root/b/doc#t{}", target);
            u.add_event(UpdateEvent::AddEdge { source_node: source_name.to_string(), 
                                               target_node: target_name.to_string(), 
                                               layer: dep_layer_name.to_string(), 
                                               component_type: AnnotationComponentType::Pointing.to_string(), 
                                               component_name: dep_comp_name.to_string() })?;
            u.add_event(UpdateEvent::AddEdgeLabel { source_node: source_name, 
                                                    target_node: target_name, 
                                                    layer: dep_layer_name.to_string(), 
                                                    component_type: AnnotationComponentType::Pointing.to_string(), 
                                                    component_name: dep_comp_name.to_string(), 
                                                    anno_ns: "".to_string(), 
                                                    anno_name: deprel_name.to_string(), 
                                                    anno_value: label.to_string() })?;
        }        
        g.apply_update(&mut u, |_msg| {})?;
        Ok(g)
    }

    fn expected_output_graph(on_disk: bool) -> Result<AnnotationGraph> {
        let mut g = AnnotationGraph::new(on_disk)?;
        let mut u = GraphUpdate::default();
        u.add_event(UpdateEvent::AddNode { node_name: "root".to_string(), node_type: "corpus".to_string() })?;        
        u.add_event(UpdateEvent::AddNode { node_name: "root/b".to_string(), node_type: "corpus".to_string() })?;
        u.add_event(UpdateEvent::AddEdge { source_node: "root/b".to_string(), 
                                           target_node: "root".to_string(), 
                                           layer: ANNIS_NS.to_string(), 
                                           component_type: AnnotationComponentType::PartOf.to_string(), 
                                           component_name: "".to_string() })?;
        u.add_event(UpdateEvent::AddNode { node_name: "root/b/doc".to_string(), node_type: "corpus".to_string() })?;
        u.add_event(UpdateEvent::AddEdge { source_node: "root/b/doc".to_string(), 
                                           target_node: "root/b".to_string(), 
                                           layer: ANNIS_NS.to_string(), 
                                           component_type: AnnotationComponentType::PartOf.to_string(), 
                                           component_name: "".to_string() })?;
        for (ii, (txt, lemma_label)) in [("I", "I"),
                                                  ("am", "be"), 
                                                  ("in", "in"), 
                                                  ("Berlin", "Berlin")].iter().enumerate() {
            let i = ii + 1;
            let name = format!("root/b/doc#t{}", i);
            u.add_event(UpdateEvent::AddNode { node_name: name.to_string(), node_type: "node".to_string() })?;
            u.add_event(UpdateEvent::AddNodeLabel { node_name: name.to_string(), 
                                                    anno_ns: ANNIS_NS.to_string(), 
                                                    anno_name: "tok".to_string(), 
                                                    anno_value: txt.to_string() })?;
            u.add_event(UpdateEvent::AddNodeLabel { node_name: name.to_string(), 
                                                    anno_ns: "".to_string(), 
                                                    anno_name: "text".to_string(), 
                                                    anno_value: txt.to_string() })?;
            u.add_event(UpdateEvent::AddNodeLabel { node_name: name.to_string(), 
                                                    anno_ns: "".to_string(), 
                                                    anno_name: "lemma".to_string(), 
                                                    anno_value: lemma_label.to_string() })?;
            if i > 1 {
                u.add_event(UpdateEvent::AddEdge { source_node: format!("root/b/doc#t{}", i - 1), 
                                                   target_node: name.to_string(), 
                                                   layer: ANNIS_NS.to_string(), 
                                                   component_type: AnnotationComponentType::Ordering.to_string(), 
                                                   component_name: "".to_string() })?;
                u.add_event(UpdateEvent::AddEdge { source_node: format!("root/b/doc#t{}", i - 1), 
                                                   target_node: name.to_string(), 
                                                   layer: ANNIS_NS.to_string(), 
                                                   component_type: AnnotationComponentType::Ordering.to_string(), 
                                                   component_name: "text".to_string() })?;
            }
        }
        let dep_layer_name = "syntax";
        let dep_comp_name = "dep";
        for (source, target) in [(2, 1), (2, 3), (3, 4)].iter() {
            let source_name = format!("root/b/doc#t{}", source);
            let target_name = format!("root/b/doc#t{}", target);
            u.add_event(UpdateEvent::AddEdge { source_node: source_name.to_string(), 
                                               target_node: target_name.to_string(), 
                                               layer: dep_layer_name.to_string(), 
                                               component_type: AnnotationComponentType::Pointing.to_string(), 
                                               component_name: dep_comp_name.to_string() })?;
        }
        g.apply_update(&mut u, |_msg| {})?;
        Ok(g)
    }

}