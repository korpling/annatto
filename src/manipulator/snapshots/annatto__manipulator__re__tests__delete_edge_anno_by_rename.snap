---
source: src/manipulator/re.rs
expression: actual.unwrap()
---
---- root.graphml:
<?xml version="1.0" encoding="UTF-8"?>
<graphml>
    <key id="k0" for="graph" attr.name="configuration" attr.type="string"/>
    <key id="k1" for="node" attr.name="lemma" attr.type="string"/>
    <key id="k2" for="node" attr.name="annis::node_type" attr.type="string"/>
    <key id="k3" for="node" attr.name="pos" attr.type="string"/>
    <key id="k4" for="node" attr.name="text" attr.type="string"/>
    <key id="k5" for="node" attr.name="annis::tok" attr.type="string"/>
    <graph edgedefault="directed" parse.order="nodesfirst" parse.nodeids="free" parse.edgeids="canonical">
        <data key="k0"><![CDATA[
# configure visualizations here
]]></data>
        <node id="root">
            <data key="k2">corpus</data>
        </node>
        <node id="root/b">
            <data key="k2">corpus</data>
        </node>
        <node id="root/b/doc">
            <data key="k2">corpus</data>
        </node>
        <node id="root/b/doc#t1">
            <data key="k1">I</data>
            <data key="k2">node</data>
            <data key="k3">PRON</data>
            <data key="k4">I</data>
            <data key="k5">I</data>
        </node>
        <node id="root/b/doc#t2">
            <data key="k1">be</data>
            <data key="k2">node</data>
            <data key="k3">VERB</data>
            <data key="k4">am</data>
            <data key="k5">am</data>
        </node>
        <node id="root/b/doc#t3">
            <data key="k1">in</data>
            <data key="k2">node</data>
            <data key="k3">ADP</data>
            <data key="k4">in</data>
            <data key="k5">in</data>
        </node>
        <node id="root/b/doc#t4">
            <data key="k1">Berlin</data>
            <data key="k2">node</data>
            <data key="k3">PROPN</data>
            <data key="k4">Berlin</data>
            <data key="k5">Berlin</data>
        </node>
        <edge id="e0" source="root/b/doc#t2" target="root/b/doc#t1" label="Pointing/syntax/dep">
        </edge>
        <edge id="e1" source="root/b/doc#t2" target="root/b/doc#t3" label="Pointing/syntax/dep">
        </edge>
        <edge id="e2" source="root/b/doc#t3" target="root/b/doc#t4" label="Pointing/syntax/dep">
        </edge>
        <edge id="e3" source="root/b/doc#t1" target="root/b/doc#t2" label="Ordering/annis/">
        </edge>
        <edge id="e4" source="root/b/doc#t2" target="root/b/doc#t3" label="Ordering/annis/">
        </edge>
        <edge id="e5" source="root/b/doc#t3" target="root/b/doc#t4" label="Ordering/annis/">
        </edge>
        <edge id="e6" source="root/b/doc#t1" target="root/b/doc#t2" label="Ordering/annis/text">
        </edge>
        <edge id="e7" source="root/b/doc#t2" target="root/b/doc#t3" label="Ordering/annis/text">
        </edge>
        <edge id="e8" source="root/b/doc#t3" target="root/b/doc#t4" label="Ordering/annis/text">
        </edge>
        <edge id="e9" source="root/b" target="root" label="PartOf/annis/">
        </edge>
        <edge id="e10" source="root/b/doc" target="root/b" label="PartOf/annis/">
        </edge>
        <edge id="e11" source="root/b/doc#t1" target="root/b/doc" label="PartOf/annis/">
        </edge>
        <edge id="e12" source="root/b/doc#t2" target="root/b/doc" label="PartOf/annis/">
        </edge>
        <edge id="e13" source="root/b/doc#t3" target="root/b/doc" label="PartOf/annis/">
        </edge>
        <edge id="e14" source="root/b/doc#t4" target="root/b/doc" label="PartOf/annis/">
        </edge>
    </graph>
</graphml>
